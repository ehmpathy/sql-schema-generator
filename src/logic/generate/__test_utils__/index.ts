export * from './databaseConnection';
export * from './createTablesForEntity';
export * from './dropTablesForEntity';
