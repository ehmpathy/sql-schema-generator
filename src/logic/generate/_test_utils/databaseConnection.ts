import mysql, { Connection, ConnectionOptions } from 'mysql2/promise';

export const getDatabaseConnection = async (): Promise<Connection> => {
  const dbConfig: ConnectionOptions = {
    // hardocded since this is just the integration test db config; settings configured in the provisioning/.../docker.file
    host: 'localhost',
    user: 'root',
    password: 'a-secure-password', // its just a docker integration test db, no worries about hardcoding this
    database: 'superimportantdb',
    port: 12821,
  };
  const connection = await mysql.createConnection(dbConfig);
  return connection;
};

export { Connection as DatabaseConnection };
