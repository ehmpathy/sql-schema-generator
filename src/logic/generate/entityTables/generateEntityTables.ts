import { Entity, Property } from '../../../domain';
import { pickKeysFromObject } from '../utils/pickKeysFromObject';
import { generateMappingTablesForArrayProperties } from './generateMappingTablesForArrayProperties';
import { generateTableForCurrentVersionPointer } from './generateTableForCurrentVersionPointer';
import { generateTableForStaticProperties } from './generateTableForStaticProperties';
import { generateTableForUpdateableProperties } from './generateTableForUpdateableProperties';

export const generateEntityTables = ({ entity }: { entity: Entity }) => {
  // 1. separate static -vs- updatable properties, pick out array properties
  const staticProps = pickKeysFromObject({
    object: entity.properties,
    keep: (property: Property) => !property.updatable,
  });
  const updatableProps = pickKeysFromObject({
    object: entity.properties,
    keep: (property: Property) => !!property.updatable,
  });
  const arrayProps = pickKeysFromObject({
    object: entity.properties,
    keep: (property: Property) => !!property.array,
  });

  // 2. validate the props
  const updatableUniqueKeys = Object.keys(updatableProps).filter((key) =>
    entity.unique.includes(key),
  );
  if (updatableUniqueKeys.length)
    throw new Error(
      `Detected a unique key on '${
        entity.name
      }' which references an updatable property. This is not supported. Keys ${JSON.stringify(
        updatableUniqueKeys,
      )}`,
    );

  // 3. define the sql
  const entityTable = generateTableForStaticProperties({
    entityName: entity.name,
    unique: entity.unique,
    properties: staticProps,
  });
  const entityVersionTable = Object.keys(updatableProps).length
    ? generateTableForUpdateableProperties({
        entityName: entity.name,
        properties: updatableProps,
      })
    : undefined;
  const entityCurrentVersionPointerTable = entityVersionTable
    ? generateTableForCurrentVersionPointer({ entityName: entity.name })
    : undefined;
  const mappingTables = generateMappingTablesForArrayProperties({
    entityName: entity.name,
    properties: arrayProps,
  });

  // 4. remove the string
  return {
    static: entityTable,
    version: entityVersionTable,
    currentVersionPointer: entityCurrentVersionPointerTable,
    mappings: mappingTables,
  };
};
