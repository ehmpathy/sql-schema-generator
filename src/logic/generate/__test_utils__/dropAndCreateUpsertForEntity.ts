import { Entity } from '../../../types';
import { generateEntityUpsert } from '../entityFunctions/generateEntityUpsert';
import { DatabaseConnection } from './databaseConnection';
import { provisionGetFromDelimiterSplitStringFunction } from './provisionGetFromDelimiterSplitStringFunction';

export const dropAndCreateUpsertFunctionForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  // if has array properties, provision the array access function
  if (Object.values(entity.properties).some((prop) => !!prop.array)) {
    await provisionGetFromDelimiterSplitStringFunction({ dbConnection });
  }

  // drop and create the upsert
  const { name, sql: upsertSql } = generateEntityUpsert({ entity });
  await dbConnection.query({ sql: `DROP FUNCTION IF EXISTS ${name}` });
  await dbConnection.query({ sql: upsertSql });
};
