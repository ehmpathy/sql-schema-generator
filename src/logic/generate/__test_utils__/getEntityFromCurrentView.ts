import { Entity } from '../../../types';
import { generateEntityCurrentView } from '../entityViews/generateEntityCurrentView';
import { DatabaseConnection } from './databaseConnection';

export const getEntityFromCurrentView = async ({
  id,
  entity,
  dbConnection,
}: {
  id: number;
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const view = generateEntityCurrentView({ entity });
  const name = view!.name;
  const results = (await dbConnection.execute({ sql: `select * from ${name} where id = ${id}` })) as any;
  expect(results[0].length).toEqual(1);
  const entityData = results[0][0];
  return entityData;
};
