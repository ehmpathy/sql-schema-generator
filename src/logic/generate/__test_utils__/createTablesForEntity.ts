import { Entity } from '../../../types';
import { generateEntityTables } from '../entityTables/generateEntityTables';
import { DatabaseConnection } from './databaseConnection';

export const createTablesForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const tables = await generateEntityTables({ entity });

  // create the static table
  await dbConnection.query({ sql: tables.static.sql });

  // drop the version table, if exists
  if (tables.version) {
    await dbConnection.query({ sql: tables.version.sql });
  }

  // drop the cvp table, if exists
  if (tables.currentVersionPointer) {
    await dbConnection.query({ sql: tables.currentVersionPointer.sql });
  }

  // create each mapping table
  await Promise.all(tables.mappings.map((mappingTable) => dbConnection.query({ sql: mappingTable.sql })));
};
