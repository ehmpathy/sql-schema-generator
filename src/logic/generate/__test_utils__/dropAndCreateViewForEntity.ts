import { DatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { Entity } from '../../../types';
import { generateEntityCurrentView } from '../entityViews/generateEntityCurrentView';

export const dropAndCreateViewForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const { name, sql: upsertSql } = generateEntityCurrentView({ entity })!;
  await dbConnection.query({ sql: `DROP VIEW IF EXISTS ${name}` });
  await dbConnection.query({ sql: upsertSql });
  return { name, sql: upsertSql };
};
