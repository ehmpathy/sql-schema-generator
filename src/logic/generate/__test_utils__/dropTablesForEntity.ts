import { DatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { Entity } from '../../../domain';
import { generateEntityTables } from '../entityTables/generateEntityTables';

export const dropTablesForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const tables = await generateEntityTables({ entity });

  // drop each mapping table
  await Promise.all(
    tables.mappings.map((mappingTable) =>
      dbConnection.query({ sql: `DROP TABLE IF EXISTS ${mappingTable.name} CASCADE;` }),
    ),
  );

  // drop the cvp table, if exists
  if (tables.currentVersionPointer) {
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.currentVersionPointer.name} CASCADE;` });
  }

  // drop the version table, if exists
  if (tables.version) {
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.version!.name} CASCADE;` });
  }

  // drop the static table
  await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name} CASCADE;` });
};
