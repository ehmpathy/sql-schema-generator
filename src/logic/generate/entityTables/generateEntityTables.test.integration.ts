import { DatabaseConnection, getDatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { getShowCreateTable } from '../../../__test_utils__/getShowCreateTable';
import { Entity, ValueObject } from '../../../domain';
import * as prop from '../../define/defineProperty';
import { createTablesForEntity, dropTablesForEntity } from '../__test_utils__';
import { generateEntityTables } from './generateEntityTables';

/*
  we test an example of every variation against the database to ensure we are defining valid sql
*/
describe('generateEntityTables', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDatabaseConnection();
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  const getShowCreateNow = async ({ tableName }: { tableName: string }) =>
    getShowCreateTable({ table: tableName, dbConnection });
  it('generates tables for a static entity, w/ same syntax as SHOW CREATE', async () => {
    const address = new Entity({
      name: 'address',
      properties: {
        street: prop.VARCHAR(255),
        city: prop.VARCHAR(255),
        country: prop.ENUM(['US', 'CA', 'MX']),
      },
      unique: ['street', 'city', 'country'],
    });
    const tables = await generateEntityTables({ entity: address });
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name};` });
    await dbConnection.query({ sql: tables.static.sql });
    expect(tables.version).toEqual(undefined);

    // check syntax is the same as that returned by SHOW CREATE TABLE
    const createStaticSql = await getShowCreateNow({ tableName: tables.static.name });
    expect(createStaticSql).toEqual(tables.static.sql); // should be the exact string
  });
  it('generates tables for a static entity that is unique on uuid, w/ same syntax as SHOW CREATE', async () => {
    const order = new Entity({
      name: 'purchase_order',
      properties: {
        value: prop.NUMERIC(5, 2),
        customer_id: prop.BIGINT(), // should be references, but to make it simple
      },
      unique: ['uuid'], // unique on uuid because same order can be placed many different times, so uuid is the only unique attribute
    });
    const tables = await generateEntityTables({ entity: order });
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name};` });
    await dbConnection.query({ sql: tables.static.sql });
    expect(tables.version).toEqual(undefined);

    // check syntax is the same as that returned by SHOW CREATE TABLE
    const createStaticSql = await getShowCreateNow({ tableName: tables.static.name });
    expect(createStaticSql).toEqual(tables.static.sql); // should be the exact string
  });
  it('generates tables for a versioned entity, w/ same syntax as SHOW CREATE', async () => {
    const contractor = new Entity({
      name: 'contractor',
      properties: {
        cognito_uuid: prop.UUID(),
        name: {
          ...prop.VARCHAR(255),
          updatable: true,
        },
        bio: {
          ...prop.TEXT(),
          updatable: true,
          nullable: true,
        },
      },
      unique: ['cognito_uuid'],
    });
    const tables = await generateEntityTables({ entity: contractor });
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.currentVersionPointer!.name};` });
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.version!.name};` });
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name};` });
    await dbConnection.query({ sql: tables.static.sql });
    await dbConnection.query({ sql: tables.version!.sql });
    await dbConnection.query({ sql: tables.currentVersionPointer!.sql });

    // check syntax is the same as that returned by SHOW CREATE TABLE
    const createStaticSql = await getShowCreateNow({ tableName: tables.static.name });
    expect(createStaticSql).toEqual(tables.static.sql); // should be the exact string
    const createVersionSql = await getShowCreateNow({ tableName: tables.version!.name });
    expect(createVersionSql).toEqual(tables.version!.sql); // should be the exact string
    const createCurrentVersionPointerSql = await getShowCreateNow({ tableName: tables.currentVersionPointer!.name });
    expect(createCurrentVersionPointerSql).toEqual(tables.currentVersionPointer!.sql); // should be the exact string
  });
  it('generates tables for an entity that references other entities - including by version', async () => {
    const wikiUser = new Entity({
      name: 'wiki_user',
      properties: {
        name: prop.VARCHAR(255),
        phone_number: prop.VARCHAR(255),
      },
      unique: ['phone_number'], // for this example, lets assume users cant change phone numbers
    });
    const wikiPage = new Entity({
      name: 'wiki_page',
      properties: {
        title: { ...prop.VARCHAR(255), updatable: true },
        content: { ...prop.TEXT(), updatable: true },
      },
      unique: ['uuid'], // unique on uuid as everything is updatable, so no natural keys
    });
    const wikiEdit = new Entity({
      name: 'wiki_edit',
      properties: {
        wiki_page_id: prop.REFERENCES(wikiPage),
        editor_id: prop.REFERENCES(wikiUser),
        final_text: { ...prop.VARCHAR(255), updatable: true },
        status: { ...prop.ENUM(['APPROVED', 'EDITABLE', 'REJECTED']), nullable: true, updatable: true },
        produced_wiki_page_version: { ...prop.REFERENCES_VERSION(wikiPage), nullable: true, updatable: true }, // i.e., what version it produced
      },
      unique: ['uuid'], // unique on uuid alone since the same user can edit the same page many times, so no natural keys
    });

    // provision the actual tables for the entity
    await dropTablesForEntity({ entity: wikiEdit, dbConnection });
    await dropTablesForEntity({ entity: wikiPage, dbConnection });
    await dropTablesForEntity({ entity: wikiUser, dbConnection });
    await createTablesForEntity({ entity: wikiUser, dbConnection });
    await createTablesForEntity({ entity: wikiPage, dbConnection });
    await createTablesForEntity({ entity: wikiEdit, dbConnection });

    // check syntax is the same as that returned by SHOW CREATE TABLE
    const tables = await generateEntityTables({ entity: wikiEdit });
    const createStaticTable = await getShowCreateNow({ tableName: tables.static.name });
    expect(createStaticTable).toEqual(tables.static.sql); // should be the exact string
    const createVersionTable = await getShowCreateNow({ tableName: tables.version!.name });
    expect(createVersionTable).toEqual(tables.version!.sql); // should be the exact string
  });
  it('generates tables for an entity with array properties (both updatable and static) and unique on one array, w/ the same syntax as show create', async () => {
    const photo = new ValueObject({
      name: 'photo',
      properties: {
        url: prop.VARCHAR(255),
      },
    });
    const host = new ValueObject({
      name: 'host',
      properties: {
        name: prop.VARCHAR(255),
      },
    });
    const home = new Entity({
      name: 'home',
      properties: {
        name: prop.VARCHAR(255),
        host_ids: prop.ARRAY_OF(prop.REFERENCES(host)),
        photo_ids: {
          ...prop.ARRAY_OF(prop.REFERENCES(photo)),
          updatable: true, // the photos of a home change over time
        },
      },
      unique: ['name', 'host_ids'],
    });

    // provision the actual tables for the entity
    await dropTablesForEntity({ entity: home, dbConnection });
    await dropTablesForEntity({ entity: host, dbConnection });
    await dropTablesForEntity({ entity: photo, dbConnection });
    await createTablesForEntity({ entity: photo, dbConnection });
    await createTablesForEntity({ entity: host, dbConnection });
    await createTablesForEntity({ entity: home, dbConnection });

    // check syntax is the same as that returned by SHOW CREATE TABLE, for each mapping table
    const tables = await generateEntityTables({ entity: home });
    expect(tables.mappings.length).toEqual(2); // two mapping tables, since two array properties
    const createMappingTableSqlOne = await getShowCreateNow({ tableName: tables.mappings[0]!.name });
    expect(createMappingTableSqlOne).toEqual(tables.mappings[0]!.sql); // should be the exact string
    const createMappingTableSqlTwo = await getShowCreateNow({ tableName: tables.mappings[1]!.name });
    expect(createMappingTableSqlTwo).toEqual(tables.mappings[1]!.sql); // should be the exact string

    // record a snapshot to confirm aesthetic acceptability
    expect(tables).toMatchSnapshot();
  });
});
