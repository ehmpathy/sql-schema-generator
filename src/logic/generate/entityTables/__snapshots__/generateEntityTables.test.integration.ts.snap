// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEntityTables generates tables for an entity with array properties (both updatable and static) and unique on one array, w/ the same syntax as show create 1`] = `
Object {
  "currentVersionPointer": Object {
    "name": "home_cvp",
    "sql": "CREATE TABLE home_cvp (
  id bigserial NOT NULL,
  updated_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  home_version_id bigint NOT NULL,
  CONSTRAINT home_cvp_pk PRIMARY KEY (id),
  CONSTRAINT home_cvp_ux1 UNIQUE (home_id),
  CONSTRAINT home_cvp_fk0 FOREIGN KEY (home_id) REFERENCES home (id),
  CONSTRAINT home_cvp_fk1 FOREIGN KEY (home_version_id) REFERENCES home_version (id)
);
CREATE INDEX home_cvp_fk0_ix ON home_cvp USING btree (home_id);
CREATE INDEX home_cvp_fk1_ix ON home_cvp USING btree (home_version_id);",
  },
  "mappings": Array [
    Object {
      "name": "home_to_host",
      "sql": "CREATE TABLE home_to_host (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_id bigint NOT NULL,
  host_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_to_host_pk PRIMARY KEY (id),
  CONSTRAINT home_to_host_ux1 UNIQUE (home_id, array_order_index),
  CONSTRAINT home_to_host_fk0 FOREIGN KEY (home_id) REFERENCES home (id),
  CONSTRAINT home_to_host_fk1 FOREIGN KEY (host_id) REFERENCES host (id)
);
CREATE INDEX home_to_host_fk0_ix ON home_to_host USING btree (home_id);
CREATE INDEX home_to_host_fk1_ix ON home_to_host USING btree (host_id);",
    },
    Object {
      "name": "home_version_to_photo",
      "sql": "CREATE TABLE home_version_to_photo (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  home_version_id bigint NOT NULL,
  photo_id bigint NOT NULL,
  array_order_index smallint NOT NULL,
  CONSTRAINT home_version_to_photo_pk PRIMARY KEY (id),
  CONSTRAINT home_version_to_photo_ux1 UNIQUE (home_version_id, array_order_index),
  CONSTRAINT home_version_to_photo_fk0 FOREIGN KEY (home_version_id) REFERENCES home_version (id),
  CONSTRAINT home_version_to_photo_fk1 FOREIGN KEY (photo_id) REFERENCES photo (id)
);
CREATE INDEX home_version_to_photo_fk0_ix ON home_version_to_photo USING btree (home_version_id);
CREATE INDEX home_version_to_photo_fk1_ix ON home_version_to_photo USING btree (photo_id);",
    },
  ],
  "static": Object {
    "name": "home",
    "sql": "CREATE TABLE home (
  id bigserial NOT NULL,
  uuid uuid NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  name varchar(255) NOT NULL,
  host_ids_hash bytea NOT NULL,
  CONSTRAINT home_pk PRIMARY KEY (id),
  CONSTRAINT home_ux1 UNIQUE (name, host_ids_hash)
);",
  },
  "version": Object {
    "name": "home_version",
    "sql": "CREATE TABLE home_version (
  id bigserial NOT NULL,
  home_id bigint NOT NULL,
  effective_at timestamp with time zone NOT NULL DEFAULT now(),
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  photo_ids_hash bytea NOT NULL,
  CONSTRAINT home_version_pk PRIMARY KEY (id),
  CONSTRAINT home_version_ux1 UNIQUE (home_id, effective_at, created_at),
  CONSTRAINT home_version_fk0 FOREIGN KEY (home_id) REFERENCES home (id)
);
CREATE INDEX home_version_fk0_ix ON home_version USING btree (home_id);",
  },
}
`;
